lib/Lift.cpp

```diff
@@ -540,59 +541,56 @@ CreateConstFromMemory(const uint64_t addr, llvm::Type *type,
   llvm::Constant *result{nullptr};
   switch (type->getTypeID()) {
     case llvm::Type::IntegerTyID: {
-      const auto size = dl.getTypeSizeInBits(type);
+      const auto size = dl.getTypeAllocSize(type);
       auto val = ReadValueFromMemory(addr, size, arch, program);
       result = llvm::ConstantInt::get(type, val);
     } break;
 
     case llvm::Type::PointerTyID: {
+      const auto pointer_type = llvm::dyn_cast<llvm::PointerType>(type);
+      const auto size = dl.getTypeAllocSize(type);
+      auto val = ReadValueFromMemory(addr, size, arch, program);
+      result = llvm::Constant::getIntegerValue(pointer_type, val);
     } break;
 
     case llvm::Type::StructTyID: {
+
       // Take apart the structure type, recursing into each element
       // so that we can create a constant structure
-      auto struct_type = llvm::dyn_cast<llvm::StructType>(type);
-
-      auto num_elms = struct_type->getNumElements();
-      auto elm_offset = 0;
-
-      std::vector<llvm::Constant *> const_list;
+      const auto struct_type = llvm::dyn_cast<llvm::StructType>(type);
+      const auto layout = dl.getStructLayout(struct_type);
+      const auto num_elms = struct_type->getStructNumElements();
+      std::vector<llvm::Constant *> initializer_list;
+      initializer_list.reserve(num_elms);
 
       for (std::uint64_t i = 0U; i < num_elms; ++i) {
-        auto elm_type = struct_type->getElementType(i);
-        auto elm_size = dl.getTypeSizeInBits(elm_type);
-
-        auto const_elm =
-            CreateConstFromMemory(addr + elm_offset, elm_type, arch,
-                                  program, module);
-
-        const_list.push_back(const_elm);
-        elm_offset += elm_size / 8;
+        const auto elm_type = struct_type->getStructElementType(i);
+        const auto offset = layout->getElementOffset(i);
+        auto const_elm = CreateConstFromMemory(addr + offset, elm_type,
+                                               arch, program, module);
+        initializer_list.push_back(const_elm);
       }
-
       result = llvm::ConstantStruct::get(struct_type,
-                                        llvm::ArrayRef(const_list));
+                                         llvm::ArrayRef(initializer_list));
```

Comment:
Is the explicit conversion needed? IIRC there should be a conversion method from `std::vector` to `llvm::ArrayRef`.